lib/plug/router.ex

```diff
@@ -334,21 +334,42 @@ defmodule Plug.Router do
         true ->
           raise ArgumentError, message: "expected :do to be given as option"
       end
-
+  
     {path, guards} = extract_path_and_guards(expr)
-
+    private_options = extract_private_options(options)
+    
     quote bind_quoted: [method: method,
                         path: path,
                         options: options,
+                        private_options: Macro.escape(private_options, unquote: true),
                         guards: Macro.escape(guards, unquote: true),
                         body: Macro.escape(body, unquote: true)] do
+
       {method, match, host, guards} = Plug.Router.__route__(method, path, guards, options)
+
       defp do_match(unquote(method), unquote(match), unquote(host)) when unquote(guards) do
-        fn var!(conn) -> unquote(body) end
+        fn conn ->
+          var!(conn) = %{ conn | private: Map.merge(conn.private, unquote(private_options)) }
+          unquote(body)
+        end
       end
     end
   end
 
+  # Pops key :private from options and checks if it is a Map. Case it is not, 
+  # raises early so that the developer knows in compilation time. Currently it
+  # does not accept options from forward calls but does not raise.
+  defp extract_private_options([]), do: {:%{}, [], []}
+  defp extract_private_options({:options, _, _}), do: {:%{}, [], []} 
```

Comment:
This is special casing the options variable and it seems to be a bug. I think the best would be to retrieve the private options inside the quoted expression (after they are evaluated) here: https://github.com/elixir-lang/plug/pull/244/files#diff-d72dd9b8aebe867306cdeee36a307a95R348

You can make `__route__` return a fifth element in the tuple (which is `private`) and then use it. :)